parlai/core/teachers.py

```diff
@@ -589,55 +589,58 @@ def _read_episode(self, data_loader):
-                else:
-                    new_entry.append(None)
-                if len(entry) > 1:
-                    # process labels if available
-                    if entry[1] is None:
-                        new_entry.append(None)
-                    elif hasattr(entry[1], '__iter__') and type(entry[1]) is not str:
-                        # make sure iterable over labels, not single string
-                        new_entry.append(tuple(sys.intern(e) for e in entry[1]))
+            if isinstance(entry, dict):
+                episode.append(entry)
+            else:
+                # intern all strings so we don't store them more than once
+                new_entry = []
+                if len(entry) > 0:
+                    # process text if available
+                    if entry[0] is not None:
+                        new_entry.append(sys.intern(entry[0]))
                     else:
-                        raise TypeError(
-                            'Must provide iterable over labels, not a single string.'
-                        )
-                    if len(entry) > 2:
-                        # process reward if available
-                        if entry[2] is not None:
-                            new_entry.append(entry[2])
-                        else:
+                        new_entry.append(None)
+                    if len(entry) > 1:
+                        # process labels if available
+                        if entry[1] is None:
                             new_entry.append(None)
-                        if len(entry) > 3:
-                            # process label candidates if available
-                            if entry[3] is None:
-                                new_entry.append(None)
-                            elif last_cands and entry[3] is last_cands:
-                                # if cands are shared, say "same" so we
-                                # don't store them again
-                                new_entry.append(
-                                    sys.intern('same as last time'))
-                            elif (hasattr(entry[3], '__iter__') and
-                                    type(entry[3]) is not str):
-                                # make sure iterable over candidates, not single string
-                                last_cands = entry[3]
-                                new_entry.append(tuple(
-                                    sys.intern(e) for e in entry[3]))
+                        elif hasattr(entry[1], '__iter__') and type(entry[1]) is not str:
+                            # make sure iterable over labels, not single string
+                            new_entry.append(tuple(sys.intern(e) for e in entry[1]))
+                        else:
+                            raise TypeError(
+                                'Must provide iterable over labels, not a single string.'
+                            )
+                        if len(entry) > 2:
```

Comment:
You can use this chance to unindent some of these right? no need to put the if X > 3 inside X > 2?